tla-io/src/main/scala/at/forsyte/apalache/io/itf/ItfToTla.scala

```diff
@@ -0,0 +1,249 @@
+
+  private[itf] def typeDrivenBuild(json: T, tt1: TlaType1): TBuilderInstruction =
+    attemptUnserializable(json).getOrElse {
+      tt1 match {
+        case BoolT1 =>
+          tla.bool(scalaFactory.asBool(json))
+        case StrT1 =>
+          val s = scalaFactory.asStr(json)
+          if (ModelValueHandler.isModelValue(s))
+            throw new JsonDeserializationError(
+                s"$s is annotated as a string, but has the value of an uninterpreted literal.")
+          tla.str(s)
+        case ct: ConstT1 =>
+          val raw = scalaFactory.asStr(json)
+          // ConstT1-type literals must adhere to the pattern regex
+          val typeIndexOpt = ModelValueHandler.typeAndIndex(raw)
+          val (constT, idx) = typeIndexOpt.getOrElse {
+            throw new JsonDeserializationError(s"$raw is annotated as an uninterpreted literal, but has string form.")
+          }
+          // The pattern-suffix must also match the declared type
+          if (ct != constT) {
+            throw new JsonDeserializationError(s"$raw is annotated as $ct but has a $constT value.")
+          }
+          tla.const(idx, constT)
+        case IntT1 =>
+          // Big integers are written as strings, not JS integers
+          if (json.allFieldsOpt.contains(Set(BIG_INT_FIELD))) {
+            val biString = scalaFactory.asStr(json.getFieldOpt(BIG_INT_FIELD).get)
+            tla.int(BigInt(biString))
+          } else tla.int(scalaFactory.asInt(json))
+
+        case SeqT1(elemT) =>
+          val elems = scalaFactory.asSeq(json)
+          val args = elems.map { typeDrivenBuild(_, elemT) }
+          // Empty collections have special builder ctors
+          if (args.isEmpty) tla.emptySeq(elemT)
+          else tla.seq(args: _*)
+
+        case RecT1(fieldTypes) =>
+          val keys = fieldTypes.keySet
+          val diffOpt = for {
+            fields <- json.allFieldsOpt
+          } yield (keys -- fields) ++ (fields -- keys) // symmetric difference
+
+          // The above is None, iff json is not an Object with fields
+          val diff = diffOpt.getOrElse {
+            throw new JsonDeserializationError(s"Record object contents are malformed.")
+          }
+
+          // The record type and value must declare the _same_ fields
+          if (diff.nonEmpty)
+            throw new JsonDeserializationError(
+                s"Record object and its type annotation $tt1 define different key sets. " +
+                  s"The keys ${diff.toSeq.mkString(", ")} are present in exactly one of the two.")
+
+          // TODO: We currently don't check the key shape against any sort of pattern.
+          // Technically, "_-?*+" is a valid record field, as far as this code is concerned.
+          // Do we want to restrict the permissible record field names, and if so, to what?
+
+          val recElems = json.allFieldsOpt.get.toSeq.map { k =>
+            k -> typeDrivenBuild(json.getFieldOpt(k).get, fieldTypes(k))
```

Comment:
Doesn't ujson provide an easy way of mapping over key/value pairs of an object? Why the need for the `option -> get -> toSeq` roundabout?

Reply:
Because this is not written for `ujson`, but for the `JsonRepresentation` abstraction.